### PR TITLE
I PVC You Pee

### PR DESCRIPTION
**Pull request by bencerr**
Used SPI to read the voltage obtained from the MAX22530, made a driver to test the registers (1 - 8).

### CODE REVIEW

**Comment by mjmagee991** [comment]
[MAX ADC Driver](https://trello.com/c/2WBigAMe/7-max-adc-driver)

File: include/PreCharge/dev/MAX22530.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#include <EVT/io/SPI.hpp>
+#include <PreCharge/PreCharge.hpp>
+#include <cstdint>
+
+namespace PreCharge {
+
+/**
+ * Handles reading the voltage values & converting it to decivolts from the
```
Comment (mjmagee991) [inline include/PreCharge/dev/MAX22530.hpp]: NIT: Goes into a bit too much detail. Doesn't need to go beyond "Handles reading voltage data from the MAX22530 ADC

File: include/PreCharge/dev/MAX22530.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * MAX22530.
+ */
+class MAX22530 {
+public:
+    /**
+     * Creates a new MAX22530 which will read a raw ADC voltage and convert it
+     * to decivolts.
```
Comment (mjmagee991) [inline include/PreCharge/dev/MAX22530.hpp:14]: NIT: There should be a newline between the summary and parameters in each method comment

File: include/PreCharge/dev/MAX22530.hpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    uint8_t readVoltage(uint16_t reg);
+
+private:
+    /// The SPI interface to read from
+    IO::SPI& spi;
+    /// Function that converts raw ADC values into decivolts
+    static uint8_t convertToVoltage(uint16_t count);
+};
+
+}// namespace PreCharge
```
Comment (mjmagee991) [inline include/PreCharge/dev/MAX22530.hpp]: Add a newline to the end of the file

File: include/PreCharge/dev/MAX22530.hpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * Returns the voltage in decivolts
+     *
+     * @return The conversion of the ADC value into temperature
+     */
+    uint8_t readVoltage(uint16_t reg);
+
+private:
+    /// The SPI interface to read from
```
Comment (mjmagee991) [inline include/PreCharge/dev/MAX22530.hpp]: These 1-line comments should follow this format:
`/** Documentation documentation */`

File: include/PreCharge/dev/MAX22530.hpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+     * Creates a new MAX22530 which will read a raw ADC voltage and convert it
+     * to decivolts.
+     * @param[in] spi
+     */
+    explicit MAX22530(IO::SPI& spi);
+
+    /**
+     * Returns the voltage in decivolts
+     *
+     * @return The conversion of the ADC value into temperature
```
Comment (mjmagee991) [inline include/PreCharge/dev/MAX22530.hpp]: This comment should be updated

File: src/PreCharge/dev/MAX22530.cpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    return result;                                       // returns in dV
+}
+
+uint8_t MAX22530::readVoltage(uint16_t reg) {
+    uint8_t bytes[2];
+    spi.readReg(0, reg, bytes, 2);
+    return convertToVoltage((bytes[0] << 8) | bytes[1]);
+}
+
+}// namespace PreCharge
```
Comment (mjmagee991) [inline src/PreCharge/dev/MAX22530.cpp]: Add a newline to the end of the file

File: src/PreCharge/dev/MAX22530.cpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#include <PreCharge/dev/MAX22530.hpp>
+
+#include <EVT/io/SPI.hpp>
+
+namespace PreCharge {
+
+MAX22530::MAX22530(IO::SPI& SPI) : spi(SPI) {}
+
+uint8_t MAX22530::convertToVoltage(uint16_t count) {     // count is out of 4096
+    uint8_t result = (((uint32_t) count) * 1098) / 40960;// 1.8 * 61 (* 10 for 1.8 to make it fix point) 1.8 is the reference voltage, 61 is voltage divider ratio
```
Comment (mjmagee991) [inline src/PreCharge/dev/MAX22530.cpp]:
```suggestion
    uint8_t result = (((uint32_t) count) * 1098) / 40960;// 1.8 * 61 (* 10 for 1.8 to make it fixed point) 1.8 is the reference voltage, 61 is voltage divider ratio
```

File: targets/MAX22530/main.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
+/**
+ * This example prints out voltage reading from the MAX22530 ADC converter
```
Comment (mjmagee991) [inline targets/MAX22530/main.cpp]:
```suggestion
 * This example prints out voltage reading from the MAX22530 ADC
```
NIT: The C in ADC stands for converter

File: targets/MAX22530/main.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    devices[0] = &IO::getGPIO<IO::Pin::SPI_CS>(EVT::core::IO::GPIO::Direction::OUTPUT);
+    devices[0]->writePin(IO::GPIO::State::HIGH);
+    IO::SPI& spi = IO::getSPI<IO::Pin::SPI_SCK, IO::Pin::SPI_MOSI, IO::Pin::SPI_MISO>(devices, deviceCount);
+    spi.configureSPI(SPI_SPEED, SPI_MODE0, SPI_MSB_FIRST);
+
+    PreCharge::MAX22530 MAX(spi);
+
+    while (1) {
+        for (int reg = 0x01; reg <= 0x04; reg++) {
+            uart.printf("Register 0x%x: %d\r\n", reg, MAX.readVoltage(reg << 2));
```
Comment (mjmagee991) [inline targets/MAX22530/main.cpp]: Put this `<< 2` inside the `readVoltage()` code, so the user doesn't have to left shift every single time they call this function.

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
There's one little issue with the driver, and a couple nitpicks with your documentation. Otherwise, looks pretty good

**Review by AarC10 — APPROVED** [review]
Looks good besides what Matt mentioned

**Review by ActuallyTaylor — APPROVED** [review]

**Review by S-TH3TA-13 — APPROVED** [review]

File: include/PreCharge/dev/MAX22530.hpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Handles reading voltage data from the MAX22530 ADC
+ */
+class MAX22530 {
+public:
+    /**
+     * Creates a new MAX22530 which will read a raw ADC voltage and convert it
+     * to decivolts.
+     *
+     * @param[in] spi
```
Comment (mjmagee991) [inline include/PreCharge/dev/MAX22530.hpp]: This variable should have a short description

**Review by mjmagee991 — APPROVED** [review]
Nice job

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
One small issue

**Review by mjmagee991 — APPROVED** [review]
Looks good

**Pull request merged**